Small changes and debug menu.

Hatrickek opened the pull request:
Small changes and debug menu.
-Added an ingame console for debugging similar to quake console.
-Keyboard support for imgui.
-Fixed the crash which occurs when the player is faster than chunk loading.
-Moved the menu handling to input.rs from camera.rs.
-Renamed the cursor_grab_system function to handle_menu_system.

Nhlest reviewed (changes requested):
Please consult contribution guidelines.

Nhlest reviewed (commented):
Looks good to me :+1: 

But it still requires some work to be worthy of being merged.

Hatrickek committed:
patch for the rework

Hatrickek committed:
Console update.

-Taking a parameter with a command.
-Text coloring with tracy levels.
-A sample event `PlayerSpeed` used with console to change the players moving speed.

Nhlest commented on src/main.rs:
```diff
@@ -15,12 +15,12 @@ use bevy_embedded_assets::EmbeddedAssetPlugin;
-use crate::ecs::plugins::settings::SettingsPlugin;
 use crate::ecs::plugins::rendering::ShikataganaiRendererPlugins;
-
+use crate::ecs::plugins::settings::SettingsPlugin;
```
empty line among uses and mods

Nhlest commented on src/ecs/plugins/console.rs:
```diff
@@ -0,0 +1,152 @@
+impl Plugin for ConsolePlugin {
+  fn build(&self, app: &mut App) {
+    app
+      .add_system_to_stage(CoreStage::PreUpdate, (debug_console).after(ImguiFrameSystem))
```
All other UI systems are added to CoreStage::Update (which is default if you just use .add_system) No need to have a different behavior here

Nhlest commented on src/ecs/plugins/camera.rs:
```diff
@@ -67,13 +69,27 @@ impl Plugin for CameraPlugin {
+  mut player_speed: ResMut<PlayerSpeed>,
+) {
+  for event in console_events.iter() {
+    if let ConsoleCommandEvents::PlayerSpeed { speed } = event {
```
Unit enum please
PlayerSpeed(speed)

Nhlest commented on src/ecs/plugins/camera.rs:
```diff
@@ -67,13 +69,27 @@ impl Plugin for CameraPlugin {
+      .insert_resource(PlayerSpeed(5.0))
       .add_system(movement_input_system)
-      .add_system(cursor_grab_system)
+      .add_system_to_stage(CoreStage::PostUpdate, process_speed_change_event)
```
Arrange them in chronological order
PreUpdate
Update
PostUpdate

Nhlest commented on src/ecs/systems/ui.rs:
```diff
@@ -123,7 +123,7 @@ pub fn main_menu(
       let [x2, _] = ui.window_content_region_max();
 
-      imgui::Slider::new("Sensitivity", 0.0, 2.0).build(ui, &mut mouse_sensetivity.as_mut().0);
+      imgui::Slider::new("Sensitivity", 0.0, 2.0).build(ui, &mut mouse_sensitivity.as_mut().0);
```
Please include a source vindicating this change. Merriam-Webster is applicable here.

Nhlest commented on src/ecs/plugins/imgui.rs:
```diff
@@ -148,18 +151,48 @@ fn start_frame(
-    app.add_system_to_stage(CoreStage::PreUpdate, start_frame.after(InputSystem));
+    app.add_system_to_stage(
+      CoreStage::PreUpdate,
+      start_frame.label(ImguiFrameSystem).after(InputSystem),
```
Like legit on the stack fr fr, this looks ugly for no particular reason. Just put that UI system to Update and it will work without this excessive labeling

Nhlest commented on src/ecs/plugins/console.rs:
```diff
@@ -0,0 +1,152 @@
+  debug_console_opened: ResMut<ConsoleMenuOpened>,
+  big_font: NonSend<BigFont>,
+  mut items: Local<Vec<ConsoleText>>,
+  mut console_event_writer: EventWriter<ConsoleCommandEvents>,
```
I think it's better for parallelism to split all console events in multiple different enums / event writers. An event for player speed, an event for noclip or whatever etc.

Nhlest commented on src/ecs/plugins/console.rs:
```diff
@@ -0,0 +1,152 @@
+    items.push(ct);
+  }
+  imgui::Window::new("Debug Console")
+    .position([0 as f32, 0 as f32], Condition::Once)
```
0.0

Nhlest commented on src/ecs/plugins/console.rs:
```diff
@@ -0,0 +1,152 @@
+        .draw_background(true)
+        .size([active_window.width(), 250.0])
+        .begin(ui);
+      match cwindow {
```
if let Some(_) = cwindow 
or
if cwindow.is_some

Nhlest commented on src/ecs/plugins/console.rs:
```diff
@@ -0,0 +1,152 @@
+            let item_text = &item.text;
+            let item_level = &item.level;
+
+            let token;
```
let token = match

Nhlest commented on src/ecs/plugins/console.rs:
```diff
@@ -0,0 +1,152 @@
+              &Level::DEBUG => token = ui.push_style_color(StyleColor::Text, YELLOW),
+            }
+            ui.text(item_text);
+            token.pop();
```
it will be dropped already at scope end

Nhlest commented on src/ecs/plugins/console.rs:
```diff
@@ -0,0 +1,152 @@
+          }
+        }
+      }
+      drop(cwindow);
```
surround it with { } instead of manually dropping

Nhlest commented on src/ecs/plugins/console.rs:
```diff
@@ -0,0 +1,152 @@
+        .hint("input commands")
+        .build()
+      {
+        fn get_bool_command(command: &str) -> bool {
```
what the fuck, just do == "1" inline or whatever

Nhlest replied:
This might have a reason for existence but i'd rather it be some generic bool::try_from or whatever idk.

Nhlest commented on src/ecs/plugins/console.rs:
```diff
@@ -0,0 +1,152 @@
+      {
+        fn get_bool_command(command: &str) -> bool {
+          if command == "1" {
+            return true;
```
yugin moment

Nhlest commented on src/ecs/plugins/console.rs:
```diff
@@ -0,0 +1,152 @@
+        fn get_float_command(command: &str) -> f32 {
+          return command.parse::<f32>().unwrap();
+        }
+        let command_list: [&str; 3] = ["clear", "noclip", "player_speed"];
```
What's the reason for this if you're already exhaustively matching on everything

Nhlest reviewed (changes requested):
Please fix

Nhlest commented:
I removed all handling of commands / events in favor of processing small and not-so-complex events as-in in-place (by modifying appropriate resources in that debug_console system. Events with this approach should only be used to delegate multi-stage commands with complex logic.

Nhlest commented:
Also it would be better to utilise some kind of home-brew parser combinator utility to parse those commands instead of relying on string split.
